mutex_flock.go

```diff
@@ -10,36 +10,120 @@ import (
 	"path/filepath"
 	"sync"
 	"syscall"
+	"time"
 
 	"github.com/juju/errors"
 )
 
-type mutex struct {
-	fd int
-	mu sync.Mutex
+func acquire(spec Spec, timeout <-chan time.Time) (Releaser, error) {
+	done := make(chan struct{})
+	defer close(done)
+	select {
+	case result := <-acquireFlock(spec.Name, done):
+		if result.err != nil {
+			return nil, errors.Trace(result.err)
+		}
+		return result.m, nil
+	case <-timeout:
+		return nil, ErrTimeout
+	case <-spec.Cancel:
+		return nil, ErrCancelled
+	}
 }
 
-func acquire(name string) (Releaser, error) {
-	flockName := filepath.Join(os.TempDir(), "juju-"+name)
-	fd, err := syscall.Open(flockName, syscall.O_CREAT|syscall.O_RDONLY, 0600)
-	if err != nil {
-		return nil, errors.Trace(err)
+var (
+	mu     sync.Mutex
+	active = make(map[string][]*waiter)
+)
+
+type waiter struct {
+	ch   chan<- acquireResult
+	done <-chan struct{}
+}
+
+type acquireResult struct {
+	m   Releaser
+	err error
+}
+
+// acquireFlock returns a channel on which an acquireResult will be
+// sent when the mutex acquisition completes, successfully or not.
+//
+// In Go, there is no natural way to cancel a flock syscall, so we
+// allow a flock syscall to continue on even when there is nothing
+// waiting for it, and then release the lock in that case. To avoid
+// an unbounded collection of goroutines, we ensure that there is
+// only one goroutine making a flock syscall at a time, per mutex
+// name.
+func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
+	result := make(chan acquireResult)
+	w := &waiter{result, done}
+
+	mu.Lock()
+	defer mu.Unlock()
+	if waiters, ok := active[name]; ok {
+		active[name] = append(waiters, w)
+		return result
 	}
 
-	err = syscall.Flock(fd, syscall.LOCK_EX|syscall.LOCK_NB)
-	if err != nil {
-		syscall.Close(fd)
+	flockName := filepath.Join(os.TempDir(), "juju-"+name)
+	flock := func() (Releaser, error) {
+		fd, err := syscall.Open(flockName, syscall.O_CREAT|syscall.O_RDONLY, 0600)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if err := syscall.Flock(fd, syscall.LOCK_EX); err != nil {
+			syscall.Close(fd)
+			return nil, errors.Trace(err)
+		}
+		return &mutex{fd: fd}, nil
+	}
+	acquire := func() bool {
+		releaser, err := flock()
+		result := acquireResult{releaser, err}
+		mu.Lock()
+		defer mu.Unlock()
 
-		if err == syscall.EWOULDBLOCK {
-			return nil, errLocked
+		var sent bool
+		waiters := active[name]
+		for !sent && len(waiters) > 0 {
+			w, waiters = waiters[0], waiters[1:]
+			select {
+			case w.ch <- result:
+				sent = true
+			case <-w.done:
```

Comment:
Is this done equivalent to Cancel in the other lock code? Is there a reason to have the attributes named differently? Or does this loop not actually handle when we aren't looking for the lock anymore?

Reply:
w.done is closed when the caller of Acquire aborts for whatever reason. That might be because it cancelled, or it might be because of a timeout, or because of a panic.

Reply:
So maybe its just that you're looping until you find a waiter that isn't already done?
In that case we are releasing the flock between waiters, so that seems good.